### ISSUE TITLE
Improvements to original Zelda levels

### ISSUE THREAD

**Issue by schrum2** [schrum2/MM-NEAT#377]
There are still several Zelda features we do not implement ... mainly the details involving the various crazy items in the game. We also don't have bosses. As a result, some dungeons seem easier to beat than they should be. This isn't necessarily a problem, but I want us to at least be aware of exactly what is missing from each level that affects the progression through the levels.

For example, in the first dungeon, what was the item that you could get for completing that block puzzle, and why was it needed to beat the level?

Notes on dungeons 1-9 with these details should be tracked in this issue thread. We may find one or two more features we need to implement.

**Comment by schrum2** [comment]
Some notes on dungeon 1: https://www.youtube.com/watch?v=YvxNj0ak8rc
In the original game, the far left room is actually accessed via a block puzzle, not a bombable wall. However, this room is also completely pointless, because in the original game it has an old man that gives a hint that has nothing to do with the current dungeon. Should we remove the room entirely? Or leave it as a red herring?

The item in the first dungeon is the bow, which is not something we can implement. It is also not relevant to the level.

There is a map in the level (I think all dungeons have one), and I think that we should restrict display of the on-screen map until after the players find a map item. However, including such an item in GAN levels will require adding a map to the grammar.

Lastly, I think we should remove the etching on the walls that identifies bombable walls. This increases the need for exploration, and makes finding a map more important.

**Comment by schrum2** [comment]
Dungeon 2 seems mostly fine. 
https://www.youtube.com/watch?v=HgYr_jf2Vsw
There is an old man hint room that tells you something about the boss, but I think it is ok to leave it in since it has a bomable wall exit.

I wonder if we should try to make a boss enemy (capital E) that is just an enemy with large health, and whose attacks knock the player back or something else mildly interesting. It could be fun.

**Comment by schrum2** [comment]
Dungeon 3:
https://www.youtube.com/watch?v=2oyGEc1NRig
This one has the raft in it, but that is irrelevant.

Our implementation currently makes the room heading to the top of the map blocked by a soft lock, but this is actually supposed to be a push block puzzle. This eventually led to a key. I was actually able to beat the level without going there, but I still want this block puzzle implemented.

Unfortunately, we'll probably end up excluding this level from the study simply because it is shaped like a swastika, albeit the Asian variety. Still, don't want to offend anyone.

**Comment by schrum2** [comment]
Dungeon 4:
When I run `Zelda_Study.bat 3` to play the 4th dungeon, it loads the 1st dungeon! This definitely needs to be fixed @jakeryang 

Can't really comment on the actual dungeon 4 too much until this is fixed.

**Comment by schrum2** [comment]
The dungeons seem to wrap after the third one. How can I test drive 4-8?

**Comment by jakeryang** [comment]
Now able to play dungeons 4 - 8 from command line.

**Comment by schrum2** [comment]
Dungeon 4: https://www.youtube.com/watch?v=XSfWDlwgkD8
For the room one up and one to the right of the entrance, we have a key appearing in the room, but the play through shows a compass there instead. Do we even need an item here at all? The room was also dark, and was lit by using the candle. This item is probably too complicated to bother with, but it is another thing we are missing.

There are other dark rooms in this dungeon. Also, many rooms seem to have many more enemies than we are placing in the rooms ... I wonder if our rooms would be more interesting with bigger swarms of enemies? Basically, can we make the threat of death at least possible?

In the room before the one that contains the ladder (on the right), the door is supposed to be soft locked, but we have it as key locked. Definitely change this. In fact, I ran out of keys and got stuck when playing this level, and I think this is why.

For the room with the ladder: currently, the raft is simply sitting in the corner. However, in the original game, there was a block puzzle that led to a room that had the ladder in it. We should not bother with the extra room, but could we keep the block puzzle? The player has to press the block puzzle to make the ladder appear ... it would be cool.

For one of the rooms close to the exit in the upper right, we have a soft lock where there is supposed to be a push-block puzzle.

Even though we don't have bosses, we should at least make the second to last room be an enemy room with a soft lock.

**Comment by schrum2** [comment]
Dungeon 5: https://www.youtube.com/watch?v=0jDNB7tIiEk
It seems odd that the very first room has a key in it ... this is not the case in the original game.

So, this is the first dungeon which has one of those underground passages that leads to a different part of the dungeon. You implemented this by adding a southern door that jumps across the maze ... it works, though this is one of the few times when having a staircase in the center of the room would have made sense. Can we at least make that door at the bottom only reveal itself (or perhaps unlock) when a block is pushed (block puzzle room?). We should probably have a separate door color for doors that are block-puzzle locked. Maybe orange?

There is another room with an underground staircase from a block puzzle that leads to a flute. This is something we can do without ... though I'm not sure why you made the room a key room.

We currently have the original boss room with nothing in it. Make this a soft locked enemy room (it is on the left).

There seems to be an inaccessible room on the upper left that is not part of the dungeon in the original game.

**Comment by schrum2** [comment]
Can't test Dungeon 6: Executing `Zelda_Study.bat 5` is loading dungeon 5 again (the same one loaded from the command `Zelda_Study.bat 4`)

**Comment by jakeryang** [comment]
> 
> 
> Can't test Dungeon 6: Executing `Zelda_Study.bat 5` is loading dungeon 5 again (the same one loaded from the command `Zelda_Study.bat 4`)

This is a typo in the code. I fixed it, but if you don't feel like recompiling the jar, just use `Zelda_Study.bat 6` to load dungeon 6. (There was a repeat of dungeon 5)

**Comment by schrum2** [comment]
Dungeon 6: https://www.youtube.com/watch?v=B7oddoaRqy4
There are not enough keys here. If the player uses the key on the right to go up into the dead end room with the old man, then the player is stuck. Oddly, in this video, the player seems to start with 4 keys. So, to solve this problem, make the old-man room be an enemy/key room.

There is a room in here with lots of those moving spike block traps. They've also shown up in previous levels ... it's a shame we can't do anything with those ... or can we?

When moving up the left edge of the dungeon (where the tail meets the body) that room is supposed to be puzzle locked, but instead is a key room.

Also, the room above this is supposed to have its bottom door soft lock after the player enters the room.

In the room above that one (which is a mini boss room in the original game) the bottom door soft locks as it is supposed to, but the right door does not.

In the room to the right of this one, the left door soft locks, but it should not. I wonder if this issue is related to the one above.

In the lower right tip of the left side of the dungeon is a room that in the original has a block puzzle that reveals stairs that teleport the player across the maze. In your version, you added a southern door that is soft locked to teleport the player. Make this door be puzzle locked instead.

There is another puzzle room in this dungeon that opens up stairs to a wand, but we can leave this out.

The final boss room (on the right side) needs to have an enemy and a soft lock (potentially an actual boss too)

**Comment by jakeryang** [comment]
> There is a room in here with lots of those moving spike block traps. They've also shown up in previous levels ... it's a shame we can't do anything with those ... or can we?
_Moving_ traps ... no. Static traps, sure. For traps, I was thinking having a tile that appears to be a floor (or something else that's somewhat noticeable), but when you're 1 block away, make it appear a trap. And if they step over it, damage them and stop them moving. I find myself just clicking the arrow keys to get somewhere fast, having these traps will make the player move a bit slower through the dungeon.

**Comment by schrum2** [comment]
Dungeon 7: https://www.youtube.com/watch?v=x3UVoYqi7jQ
When I watch the movie for this, the room to the right of the entrance is not a key room, but a bomb does appear there when all of the enemies are killed. However, Link already had 3 keys. I wonder if key drops partially depend on whether Link already has keys. I guess we should keep dropping a key here, but in general we want to make sure that the player only has enough keys ... not too many.

Zig-zagging right, up, right from the start: The top door of this room should be soft locked, but is simply open.

The room to the right of this one has a mini boss in the original.

The room two rooms above the start room: It's right door should be soft locked. The top exit in the room above this one should also be soft locked. In both of these cases, the door returning to the previous room does indeed have a soft lock ... I think that you placed the soft locks on the edges between these rooms, but in the wrong directions. In fact, the room above this room (in your version) had both doors soft locked, but neither of them are soft locked in the original, though they are locked on the other side/room.

In the upper left of this dungeon is a room with a guy that stops Link and says "Grumble Grumble." Apparently, to get past him you have to give him some meat. This is interesting ... essentially, it boils down to an item puzzle: provide a particular item to pass. Too complicated to implement, but a fun idea.

There is a puzzle room with downward stairs leading to a candle. Not an item we need.

For the room on the far right, but one row beneath the top, the top door should be soft locked (this is a mini boss room). The door on the other side/room incorrectly shows up as soft locked.

In your version, you made the room in the absolute top right have a soft locked door that teleports the player across the maze. In the original, this is caused by stairs in the floor. However, the stairs appear from a block puzzle, so we should make your soft lock door appear (or at least unlock) because of a block puzzle.

After teleporting to roughly the mid upper area of the maze, the player bombs a wall to the right and goes into a room. Your version is empty. The original version has a boss. We should at least have enemies and a soft locked exit door.

**Comment by schrum2** [comment]
Dungeon 8: https://www.youtube.com/watch?v=M_ybJ_trQhc

You should place a ladder item in the first room of this dungeon.

Room to left of entrance should have left door soft locked (adjacent room has the soft lock instead).

The room to the left of this has the puzzle block leading underground to a secret item, which in this case is a spell book? Note sure what it's needed for.

In the room two rooms above the start room, the top and left doors need to be soft locked (the rooms on the other side are soft locked instead).

In the room above this one, there should be a locked door on the left side, but instead there is no door.

In a room on the upper right which has one room above and one to the right: The top door is correctly soft locked, but the right door is not (although the door on the other side is incorrectly soft locked).

A puzzle room in the upper right has stairs that lead down to a weird key. What is this for?

After teleporting from a room on the right edge to a one on the left side, you can bomb the northern wall to get into a new room. This room is a boss room with a soft lock on top. Add an enemy and a soft lock to it.

**Comment by jakeryang** [comment]
> A puzzle room in the upper right has stairs that lead down to a weird key. What is this for?

The key is a magical key. What makes it so magic is that it can unlock any locked door... which might cause problems in dungeon 9

**Comment by schrum2** [comment]
Back to Dungeon 1: The room right before the Triforce (door on the right) should have a soft lock, but instead it is on the left door of the Triforce room (which does nothing since there are no enemies there)

**Comment by schrum2** [comment]
Also Dungeon 1: Since the puzzle block in the upper left does nothing, we should disable it entirely. We only want puzzle blocks that actually work.

**Comment by schrum2** [comment]
There are some small problems with soft locks in the last two rooms of Dungeon 2:
![image](https://user-images.githubusercontent.com/8397236/61082360-ee799480-a3ee-11e9-8652-8a4880097f64.png)
The doors with an X should not have soft locks, but the door with a circle should have a soft lock.

**Comment by schrum2** [comment]
Dungeon 4: I like the new puzzle block room. However, when playing this I ran out of keys and got stuck. However, I'm sure that A* can still beat the level because there is a way to beat the level. The problem is that there is a locked door that leads to a dead end, so if you waste a key there, you won't have enough to progress. But if you coincidentally avoid that bad door, you can get through.

I looks like I asked you to remove a key from this dungeon in my comment above, because it is a compass in the original game ... did you remove the key? If so, put it back. Otherwise, I guess we need to add a new one.

**Comment by schrum2** [comment]
Dungeon 5 seems to have an impossible puzzle lock now:
![image](https://user-images.githubusercontent.com/8397236/61085283-75317000-a3f5-11e9-8e65-eb63c708384f.png)
There are no pushable blocks in this room, but the door is puzzle locked. In fact, the door on the other size is also marked as being a puzzle door, but there are no blocks in that room either:
![image](https://user-images.githubusercontent.com/8397236/61085353-a611a500-a3f5-11e9-9b6e-75b92a1cf270.png)

**Comment by schrum2** [comment]
Also Dungeon 5: Despite the puzzle issue mentioned above, the level is still beatable. However, this is also a level that can become unbeatable if the player happens to use their keys in the wrong order.

I've noticed in the playthrough videos that Link often starts the dungeons with some keys. It's starting to seem that we simply need to add more keys in order for the levels to be beatable. However, I'm hesitant to make that design decision in the original levels. However, I've suggested a fix in a new issue.

**Comment by schrum2** [comment]
Although you are working on https://github.com/jakeryang/ZeldaDungeons now, make sure that whenever you fix the outstanding issues in this thread, they are fixed in both repos.

**Comment by schrum2** [comment]
This issue has a lot of intricate details that we are not going to focus on at this point. The interesting science is in AI generation of levels, but this issue involves a lot of specialized programming for levels from the original game ... at lot of this will become irrelevant if we ever switch to a better game engine. Therefore, this issue is not worth pursuing any more.
